src/orion/algo/robo/ablr/ablr_model.py

```diff
@@ -0,0 +1,589 @@
+
+        self.lower, self.upper = build_bounds(self.space)
+
+    def state_dict(  # type: ignore
+        self,
+        destination=None,
+        prefix="",
+        keep_vars: bool = False,
+    ) -> OrderedDict[str, Tensor | Any]:
+        state_dict: OrderedDict[str, Any] = super().state_dict(
+            destination=destination, prefix=prefix, keep_vars=keep_vars  # type: ignore
+        )
+        state_dict["rng"] = torch.random.get_rng_state()
+        state_dict["optimizer"] = self.optimizer.state_dict()
+        return state_dict
+
+    def load_state_dict(self, state_dict: dict, strict: bool = True) -> tuple:
+        optim_state: dict = state_dict.pop("optimizer")
+        self.optimizer.load_state_dict(optim_state)
+        rng: Tensor = state_dict.pop("rng")
+        torch.random.set_rng_state(rng)
+        return super().load_state_dict(state_dict, strict=strict)  # type: ignore
+
+    def set_state(self, state_dict: dict) -> None:
+        """Orion hook used to restore the state of the algorithm."""
+        self.load_state_dict(state_dict)
+
+    def seed(self, seed: int) -> None:
+        """Seed the state of the random number generator.
+
+        :param seed: Integer seed for the random number generator.
+
+        .. note:: This methods does nothing if the algorithm is deterministic.
+        """
+        # No need to do anything here really, since the ROBO_ABLR class already seeds all the
+        # pytorch RNG.
+
+    def predict(self, X_test: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
+        """Return the predictive mean and variance for the given points."""
+        y_pred_distribution = self.predict_dist(X_test)
+        return (
+            y_pred_distribution.mean.cpu().numpy(),
+            y_pred_distribution.variance.cpu().numpy(),
+        )
+
+    def predict_dist(self, x_test: Tensor | np.ndarray) -> Normal:
+        """Return the predictive distribution for the given points."""
+        with torch.no_grad():
+            if self._predict_dist is None:
+                # re-create the predictive distribution.
+                # NOTE: This gets overwritten every time a new point is observed, so we know for
+                # sure that this is always the most up-to-date predictions.
+                _, self._predict_dist = self(self.X, self.y)
+            y_pred_dist = self._predict_dist(x_test)
+        return y_pred_dist
+
+    def forward(
+        self, x_train: Tensor, y_train: Tensor
+    ) -> tuple[Tensor, Callable[[Tensor], Normal]]:
+        """Return the training loss and a function that will give the
+        predictive distribution over y for a given (un-normalized) input x_test.
+
+        NOTE: Assumes that `x_train`, `y_train` and an eventual `x_test` are
+        NOT already rescaled/normalized!
+        """
+        x_train = torch.as_tensor(x_train)
+        y_train = torch.as_tensor(y_train)
+        # TODO: Do we normalize the portion that has the contextual information?
```

Comment:
What is the contextual information?

Reply:
That's the task id, that gets added to the "X" of the quadratics in the example of the paper. It's unclear if we should normalize that portion of the "X" matrix.

Reply:
For this version of ABLR (single task) we assume all input data is real, right? I would treat the task id the same way as the rest. Otherwise I think we would need to adapt the model/kernel to the specific type of input (categorical).

Reply:
Yeah, you're right. Removed the TODO.

Reply:
I removed this leftover todo.